cmd/snap-confine/snap-confine.c

```diff
@@ -103,6 +103,42 @@ typedef struct sc_invocation {
 	bool is_normal_mode;
 } sc_invocation;
 
+/**
+ * sc_preservation_context remembers clobbered state to restore.
+ *
+ * The umask is preserved and restored to ensure consistent permissions for
+ * runtime system. The value is preserved and restored perfectly.
+ *
+ * The current working directory is preserved and possibly restored, if it
+ * exists in the runtime execution environment. The limitation on the current
+ * working directory is that not all paths that exist on the host filesystem
+ * exist in the execution environment. When the current working directory
+ * cannot be restored then a special substitute of /var/lib/snapd/void is used
+ * instead.
```

Comment:
This cwd comment probably shouldn't be in this PR. That said, what is your plan with it? I suggest considering `cwd_fd = open(".", O_DIRECTORY | O_PATH | O_CLOEXEC)` in sc_remember_preservation_context, then `fchdir(cwd_fd) ; close(cwd_fd)` in sc_restore_preservation_context (checking return codes, etc, etc).

Reply:
FWIW, I would not mind if you added the cwd stuff to this PR, but I'll leave that to your discretion.

Reply:
> This cwd comment probably shouldn't be in this PR. That said, what is your plan with it? I suggest considering `cwd_fd = open(".", O_DIRECTORY | O_PATH | O_CLOEXEC)` in sc_remember_preservation_context, then `fchdir(cwd_fd) ; close(cwd_fd)` in sc_restore_preservation_context (checking return codes, etc, etc).

Note, because of O_PATH, we don't need apparmor rules for this.